File: plot_data.R
#!/usr/bin/Rscript
library("ggplot2")
require(mgcv)
require(gridExtra)

fn <- "HadCRUT_data.txt"
cru <- read.table(url, fill=TRUE)

# drop the other cols
cru <- cru[c(1,2)]
names(cru)[1] <- "year"
names(cru)[2] <- "temp"

golden_ratio <- 1.0 / 1.6180339887
ax1 <- ggplot(cru, aes(year, temp)) +
         geom_point(size=0.8) +
         xlab("") +
         ylab(expression(Temperature~anomaly~(~degree~C)~1961-1990)) +
         geom_hline(yintercept=0.0, linetype="dashed", colour="lightgrey") +
         stat_smooth(method="gam", formula=y~s(x, k=20), size=0.7) +
         theme_bw() +
         theme(aspect.ratio=golden_ratio,
               panel.grid.major=element_blank(),
               panel.grid.minor=element_blank())

# Check if AR1 model is a better fit
g1 <- gamm(temp~s(year, k=20), data=cru)
g2 <- gamm(temp~s(year, k=20), data=cru, correlation=corARMA(form=~year, p=1))

# AR1 model is best.
anova(g1$lme, g2$lme)

gam <- predict(g1$gam, data=cru)$fit
ar1 <- predict(g2$gam, data=cru)
se <- predict(g2$gam, data=cru, se=TRUE)$se.fit
df <- with(cru, data.frame(year=seq(min(year), max(year), length=length(year)),
                           temp=ar1,
                           lcl=ar1 - 1.96 * se,
                           ucl=ar1 + 1.96 * se))
df2 <- with(cru, data.frame(year=seq(min(year), max(year), length=length(year)),
                           temp=gam))


ax2 <- ggplot() +
         geom_point(data=cru, aes(year, temp), size=0.8) +
         geom_line(data=df, aes(year, temp), size=0.8, colour="blue") +
         geom_line(data=df2, aes(year, temp), size=0.8, colour="lightgreen") +
         geom_ribbon(data=df, aes(year, ymin=lcl, ymax=ucl),
                     fill="grey", alpha=.4) +
         ylab(expression(Temperature~anomaly~(~degree~C)~1961-1990)) +
         geom_hline(yintercept=0.0, linetype="dashed", colour="lightgrey") +
         stat_smooth(method="gam", formula=y~s(x, k=20), size=0.7) +
         theme_bw() +
         theme(aspect.ratio=golden_ratio,
              panel.grid.major=element_blank(),
              panel.grid.minor=element_blank())

grid.arrange(ax1, ax2, ncol=1)

File: plot_data.py
#!/usr/bin/env python

"""
Call the R GAM function and make the same final plot in python

That's all folks.
"""
__author__ = "Martin De Kauwe"
__version__ = "1.0 (02.05.2016)"
__email__ = "mdekauwe@gmail.com"

import os
import sys
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import rpy2
from rpy2 import robjects as ro
from rpy2.robjects.packages import importr
from rpy2.robjects import pandas2ri
pandas2ri.activate()

fn = "HadCRUT_data.txt"
df = pd.read_csv(fn, sep="   ", header=None, usecols=[0,1],
                  names=["year","temp"], engine='python')

df = df[df.year != 2016]

importr("mgcv")
rdf = pandas2ri.py2ri(df)
ro.globalenv['cru'] = rdf
ro.r('nvals = length(cru$year)')
ro.r('fit=gamm(temp ~ s(year, k=20), data=cru)')

# Check if AR1 model is a better fit as each year isn't truly independent
ro.r('g1 <- gamm(temp~s(year, k=20), data=cru)')
ro.r('''g2 <- gamm(temp~s(year, k=20), data=cru,
                   correlation=corARMA(form=~year, p=1))''')

# AR1 model is best.
print ro.r('anova(g1$lme, g2$lme)')

ro.r('gam <- predict(g1$gam, data=cru)')
ro.r('ar1 <- predict(g2$gam, data=cru)')
ro.r('se <- predict(g2$gam, data=cru, se=TRUE)$se.fit')

# get objects back for plotting
lcl = ro.r('ar1 - 1.96 * se')
ucl = ro.r('ar1 + 1.96 * se')
gam = ro.r('gam')

sns.set_style("ticks")
sns.set_style({"xtick.direction": "in","ytick.direction": "in"})
golden_mean = 0.6180339887498949
width = 9
height = width * golden_mean
fig = plt.figure(figsize=(width, height))
ax1 = fig.add_subplot(111)

ax1.plot(df.year, df.temp, "ko")
ax1.plot(df.year, gam, ls="-", lw=2.5, color="royalblue")
ax1.fill_between(df.year, lcl, ucl, color="lightgrey")
ax1.set_xlabel("Year")
ax1.set_ylabel("Temperature anomaly ($^\circ$C; 1961-1990)")
fig.savefig("HadCRU4_temperature_python.pdf", bbox_inches='tight',
            pad_inches=0.1)
